Set AWS container properties from container options

manuelesimi opened the pull request:
This contribution addresses the requirements from https://github.com/nextflow-io/nextflow/issues/2282 and https://github.com/nextflow-io/nextflow/pull/2413.

With this change, the container properties of a job definition registered by Nextflow in AWS Batch are set starting from the container options defined in the Nextflow task.

**Example**

Given the following container options:
```
process sayHelloFirst {

  containerOptions '--env MYVAR2=foo2 --env MYVAR3=foo3 --memory-swap 240000 --memory-swappiness 20 --shm-size 16000000 --ulimit nofile=1280:2560 --ulimit nproc=16:32 --privileged'

}
```
the job is registered with the following properties on AWS Batch: 

<img width="1143" alt="properties" src="https://user-images.githubusercontent.com/2822686/144300446-53f814dd-cc3a-47b5-beb5-29aa462ee06b.png">

Do note that:
1. this mapping does not include properties already set by other means in Nextflow (like volumes, mount points, cpus, memory, etc.)
2. not all the docker run options can be mapped into AWS.  See [AWS doc](https://docs.aws.amazon.com/batch/latest/APIReference/API_ContainerProperties.html).

manuelesimi commented:
I just pushed a fix for tmpfs mounts. They are now correctly set in the container properties:

<img width="555" alt="tmpfs mounts" src="https://user-images.githubusercontent.com/2822686/144463476-569a9003-58b7-440b-8595-983a39869ecd.png">

pditommaso commented on plugins/nf-amazon/src/main/nextflow/cloud/aws/batch/AWSContainerOptionsMapper.groovy:
```diff
@@ -0,0 +1,170 @@
+    def options
+
+    protected AWSContainerOptionsMapper(String containerOptions) {
+        options = containerOptions.trim().split("\\s+")
```
Too weak, the container option values could container blanks and quotes. Also I think the parsing should not be done by this component, which could take instead a Map object

manuelesimi replied:
That makes sense. I also agree that the parsing should occur earlier. Let me think about it.

pditommaso replied:
There's an implementation in the commit I've linked above

manuelesimi replied:
I saw it :)

manuelesimi replied:
I saw it, thanks :)

pditommaso reviewed (commented):
Hi Manuale, this is welcome. Actually, I had gave a try some time ago but gave up when I realised container options should be managed at Job Definition level (what a design mistake ..) 

My main comment is that this is a good opportunity to manage the `containerOption` as a map. Look at my commit that I've just pushed 

https://github.com/nextflow-io/nextflow/commit/5e9c1cce8c7037be05a77f4c68f91966d7d62521

manuelesimi commented:
Great work (and fast!).
I noticed you pushed your change for the map options to another branch. How should I proceed? Merger yours in this one, wait until you merge yours into master and then merge master back into this, or else?

pditommaso commented:
I would say to merge my branch in your PR, then you can continue with that

manuelesimi commented:
@pditommaso There is a problem with your parser. The Map doesn't consider options that can be repeated multiple times (like --env or --tmpfs). Each option should be mapped to a List of values.

manuelesimi commented:
Also, '=' cannot be used as separator in the `QuoteStringTokenizer`. There are options that uses '=' in their values. E.g. `--env MY_VAR-value`.

pditommaso commented:
It's getting complicated. Should repeatable options be supported? my original plan was only to support a subset of them.

manuelesimi commented:
> It's getting complicated. Should repeatable options be supported? my original plan was only to support a subset of them.

Parsing command line options is never easy.

I made a change that seems to work:
From `'--env VAR_FOO -e VAR_FOO2=value2 --env VAR_FOO3=value3 --user nf-user'`, I get:

<img width="342" alt="list" src="https://user-images.githubusercontent.com/2822686/144489430-86289b6e-5125-494a-9675-d0fc725b4d1e.png">

Would that be OK with you?

What's probably more concerning is that we should support only the format "option value" (which is the one documented in the docker run help), not "option=value". So a value with an equal char is not separated (like in the screenshot above).

pditommaso commented:
Is basically a `Map<String,List>` ?

manuelesimi commented:
> Is basically a `Map<String,List>` ?

yes, it is.

pditommaso commented:
Maybe we should model it as a `List` of pairs, or even a custom holder class

manuelesimi commented:
I'd go for a custom holder where we can add the lookup methods we will need.

pditommaso commented:
👍

manuelesimi commented:
Question:

The method `Map getContainerOptionsMap()` in TaskConfig returns a Map. Hence, we have the following options:

1) we change the method signature to return our custom holder class. As far as I see, it is used only [here](https://github.com/nextflow-io/nextflow/blob/6adfce099ecf3c835e03d1a2eb98ea81d0540019/plugins/nf-google/src/main/nextflow/cloud/google/lifesciences/GoogleLifeSciencesTaskHandler.groovy#L349) so it would have a low impact.
2) our custom holder extends a Map (but then we need to cast back to the holder's class to access its methods, which I'd really dislike)
3) we stick to a Map<String,List>

Which one would you favor?

pditommaso commented:
I would say 1.

manuelesimi commented:
Also a combination of 1. and 2. would work well.

manuelesimi commented:
> I would say 1.

Sounds good.

pditommaso commented:
As you wish 😄

manuelesimi commented:
@pditommaso I switched to a custom handler for the option map. I also added support for long option with dashes (docker has a few we have to support like `-memory-swap ... --shm-size ...`.
I adapted the AWS container mapper accordingly, which is greatly simplified with respect to the previous version.

See what you think.

PS: I'm marking this PR as draft since I need additional tests on AWS to make sure all is working as expected after these changes.

manuelesimi commented:
I have tested again the submissions to AWS Batch with the latest changes pushed in this PR. I can confirm that all the container properties are set correctly (see the screenshots from previous comments) according to the options specified in `containerOptions`.

As I mentioned, not all the `docker run` options are supported in the AWS job definition. If this PR is approved, should we also update the documentation?

pditommaso commented on modules/nf-commons/src/main/nextflow/util/CmdLineOptionMap.groovy:
```diff
@@ -0,0 +1,63 @@
+ *
+ * @author Manuele Simi <manuele.simi@gmail.com>
+ */
+class CmdLineOptionMap {
```
Whenever possible new classes should be annotated with `@CompileStatic`

pditommaso replied:
It's missing `CmdLineOptionMapTest` or I'm wrong?

manuelesimi replied:
Indeed, it was missing. Just added.

pditommaso commented on plugins/nf-amazon/src/main/nextflow/cloud/aws/batch/AwsContainerOptionsMapper.groovy:
```diff
@@ -0,0 +1,149 @@
+ * @see <a href="https://docs.aws.amazon.com/batch/latest/APIReference/API_ContainerProperties.html">API Container Properties</a>
+ * @author Manuele Simi <manuele.simi@gmail.com>
+ */
+class AwsContainerOptionsMapper {
```
Let's add `@CompileStatic`

manuelesimi replied:
Done

pditommaso commented on modules/nf-commons/src/main/nextflow/util/CmdLineOptionMap.groovy:
```diff
@@ -0,0 +1,63 @@
+        options.size()
+    }
+
+    def getValues(String key) {
```
Does this always return a `List`? in this case worth declaring `List getValues(String key)`

manuelesimi replied:
done

pditommaso reviewed (changes requested):
It looks a nice work! Left a few comments

pditommaso commented:
> should we also update the documentation?

it would be nice!

manuelesimi commented:
> > should we also update the documentation?
> 
> it would be nice!

See https://github.com/nextflow-io/nextflow/pull/2471/commits/75c5623c0122c70a6ba431d0cbb140b19cce0102

manuelesimi commented:
> It looks a nice work! Left a few comments

Hi @pditommaso, thanks for your additional comments. Indeed they helped to improve. 

After the changes, I've tested again with AWS Batch and the job properties are diligently matched with the container options :)

pditommaso commented on plugins/nf-amazon/src/main/nextflow/cloud/aws/batch/AwsContainerOptionsMapper.groovy:
```diff
@@ -101,25 +104,33 @@ class AwsContainerOptionsMapper {
+        }
 
-        containerProperties.setLinuxParameters(params)
+        if ( atLeastOneSet )
```
Is there a side-effect setting an empty `LinuxParameters` object? 

Also, it may be better to have more functional behaviour having `checkLinuxParameters` returning the `LinuxParameters`. 

Finally `checkLinuxParameters` could be a static method to prevent creating a class for each task.

manuelesimi replied:
> Is there a side-effect setting an empty `LinuxParameters` object?
> 

It's null by default. I wanted to keep it that way if there are no parameters set. This way we have the same (super-tested) behavior previous to this PR. If no param is set the object is null.

> Also, it may be better to have more functional behaviour having `checkLinuxParameters` returning the `LinuxParameters`.

It makes sense.
> 
> Finally `checkLinuxParameters` could be a static method to prevent creating a class for each task.

Makes sense as well, I'll change it (eventually, it'd be an object for each task :) )

manuelesimi replied:
I made the entire AWSContainerMapper class static to avoid creating unnecessary objects for each task.

pditommaso commented on modules/nf-commons/src/main/nextflow/util/CmdLineHelper.groovy:
```diff
@@ -66,12 +75,12 @@ class CmdLineHelper {
+        for( int i=0; i<vals.size(); i++ ) {
+            vals[i] = vals[i] ?. split(splitter) as String
         }
 
```
Think this can be removed.

manuelesimi replied:
> Just realised that `AwsContainerOptionsMapper` was tested via indirectly the `TaskHandler`. Think it would be better to simplify this testing directly the class ie.
> 
> ```
>  when:
>  def map = CmdLineHelper.parseGnuArgs('--ulimit nofile=1280:2560 --ulimit nproc=16:32')
>  def properties = AwsContainerOptionsMapper.createContainerOpts(map) 
>  then:
>  properties.getUlimits().size() == 2
>  properties.getUlimits().get(0).toString() == '{HardLimit: 2560,Name: nofile,SoftLimit: 1280}'
>  properties.getUlimits().get(1).toString() == '{HardLimit: 32,Name: nproc,SoftLimit: 16}'
> ```
> 
> https://github.com/nextflow-io/nextflow/blob/80dbfa8e64e42b49acf5b89040b8d9e180565c28/plugins/nf-amazon/src/test/nextflow/cloud/aws/batch/AwsContainerOptionsMapperTest.groovy#L12-L156

It is also tested there. When you asked for a test, I thought you wanted a dedicated test class. I agree and will move the tests  as you suggested.

manuelesimi replied:
done, see https://github.com/nextflow-io/nextflow/pull/2471/commits/a724067fe5c8d34ee7685bdb215b212ce7446b83

pditommaso commented on modules/nf-commons/src/main/nextflow/util/CmdLineOptionMap.groovy:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    @Override
+    Hasher funnel(Hasher hasher, CacheHelper.HashMode mode) {
+        return CacheHelper.hasher(hasher, options, mode)
+    }
```
I've added the logic to make it hasheable

manuelesimi replied:
yep, definitely needed and welcomed.

pditommaso commented on plugins/nf-amazon/src/main/nextflow/cloud/aws/batch/AwsContainerOptionsMapper.groovy:
```diff
@@ -0,0 +1,159 @@
+@CompileStatic
+class AwsContainerOptionsMapper {
+
+    static ContainerProperties createContainerOpts(CmdLineOptionMap options) {
```
Made it return the `ContainerProperties` object

manuelesimi replied:
Looks more readable. Thanks.

pditommaso commented on plugins/nf-amazon/src/main/nextflow/cloud/aws/batch/AwsContainerOptionsMapper.groovy:
```diff
@@ -0,0 +1,159 @@
+
+        return atLeastOneSet ? params : null
+    }
+
```
It should be added a unit test each of the option

manuelesimi replied:
done, see https://github.com/nextflow-io/nextflow/pull/2471/commits/d1d9da011fb3e422b08ebc9b3352768f25605984

pditommaso commented on plugins/nf-amazon/src/test/nextflow/cloud/aws/batch/AwsBatchTaskHandlerTest.groovy:
```diff
@@ -463,7 +463,9 @@ class AwsBatchTaskHandlerTest extends Specification {
+        def handler = Spy(AwsBatchTaskHandler) {
+            getTask() >> Mock(TaskRun) { getConfig() >> Mock(TaskConfig)  }
+        }
         handler.executor = Mock(AwsBatchExecutor)
```
A few tests were failing

manuelesimi replied:
thanks!

pditommaso commented on plugins/nf-amazon/src/main/nextflow/cloud/aws/batch/AwsBatchTaskHandler.groovy:
```diff
@@ -485,7 +492,10 @@ class AwsBatchTaskHandler extends TaskHandler implements BatchHandler<String,Job
+        final tokens = [name, image, awscli, volumes, jobRole]
+        if( containerOpts )
+            tokens.add(containerOpts)
+        hashingTokens.addAll(tokens)
```
Add the `containerOptions` to the hashing key to force a new Job definition when container a value change

pditommaso commented:
Just realised that `AwsContainerOptionsMapper` was tested via indirectly the `TaskHandler`. Think it would be better to simplify this testing directly the class ie. 

```
 when:
 def map = CmdLineHelper.parseGnuArgs('--ulimit nofile=1280:2560 --ulimit nproc=16:32')
 def properties = AwsContainerOptionsMapper.createContainerOpts(map) 
 then:
 properties.getUlimits().size() == 2
 properties.getUlimits().get(0).toString() == '{HardLimit: 2560,Name: nofile,SoftLimit: 1280}'
 properties.getUlimits().get(1).toString() == '{HardLimit: 32,Name: nproc,SoftLimit: 16}'
```


https://github.com/nextflow-io/nextflow/blob/80dbfa8e64e42b49acf5b89040b8d9e180565c28/plugins/nf-amazon/src/test/nextflow/cloud/aws/batch/AwsContainerOptionsMapperTest.groovy#L12-L156

manuelesimi commented:
> Just realised that `AwsContainerOptionsMapper` was tested via indirectly the `TaskHandler`. Think it would be better to simplify this testing directly the class ie.
> 
> ```
>  when:
>  def map = CmdLineHelper.parseGnuArgs('--ulimit nofile=1280:2560 --ulimit nproc=16:32')
>  def properties = AwsContainerOptionsMapper.createContainerOpts(map) 
>  then:
>  properties.getUlimits().size() == 2
>  properties.getUlimits().get(0).toString() == '{HardLimit: 2560,Name: nofile,SoftLimit: 1280}'
>  properties.getUlimits().get(1).toString() == '{HardLimit: 32,Name: nproc,SoftLimit: 16}'
> ```
> 
> https://github.com/nextflow-io/nextflow/blob/80dbfa8e64e42b49acf5b89040b8d9e180565c28/plugins/nf-amazon/src/test/nextflow/cloud/aws/batch/AwsContainerOptionsMapperTest.groovy#L12-L156

It is also tested there. When you asked for a test, I thought you wanted a dedicated test class. I agree and will move the tests as you suggested.

manuelesimi commented:
> > Just realised that `AwsContainerOptionsMapper` was tested via indirectly the `TaskHandler`. Think it would be better to simplify this testing directly the class ie.
> > ```
> >  when:
> >  def map = CmdLineHelper.parseGnuArgs('--ulimit nofile=1280:2560 --ulimit nproc=16:32')
> >  def properties = AwsContainerOptionsMapper.createContainerOpts(map) 
> >  then:
> >  properties.getUlimits().size() == 2
> >  properties.getUlimits().get(0).toString() == '{HardLimit: 2560,Name: nofile,SoftLimit: 1280}'
> >  properties.getUlimits().get(1).toString() == '{HardLimit: 32,Name: nproc,SoftLimit: 16}'
> > ```
> > 
> 
> It is also tested there. When you asked for a test, I thought you wanted a dedicated test class. I agree and will adjust the tests as you suggested.

done, see https://github.com/nextflow-io/nextflow/pull/2471/commits/d1d9da011fb3e422b08ebc9b3352768f25605984

manuelesimi commented:
I think I've addressed your latest comments. Pls, let me know if I missed something. Thanks!

manuelesimi commented:
Hold on, there is a test failing. I know why. I can fix it later.

manuelesimi commented:
So, I changed two `GoogleLifeSciencesTaskHandlerTest` tests that uses the new config map. I hope I didn't misinterpret them, but basically I'd expect that when the entrypoint is not set in the map, the default entrypoint is returned. See https://github.com/nextflow-io/nextflow/pull/2471/commits/96c3fd4e79ddfd028380fc0b01aef3494450c59e

pditommaso commented on plugins/nf-google/src/test/nextflow/cloud/google/lifesciences/GoogleLifeSciencesTaskHandlerTest.groovy:
```diff
@@ -197,7 +197,7 @@ class GoogleLifeSciencesTaskHandlerTest extends GoogleSpecification {
         and:
-        req.entryPoint == null
+        req.entryPoint == GoogleLifeSciencesConfig.DEFAULT_ENTRY_POINT
 
```
Interesting, why before it was null? 😕

pditommaso replied:
So, the constant is `/bin/bash`

https://github.com/nextflow-io/nextflow/blob/96c3fd4e79ddfd028380fc0b01aef3494450c59e/plugins/nf-google/src/main/nextflow/cloud/google/lifesciences/GoogleLifeSciencesConfig.groovy#L48-L48

However when is null defaults to `bash` 

https://github.com/nextflow-io/nextflow/blob/96c3fd4e79ddfd028380fc0b01aef3494450c59e/plugins/nf-google/src/main/nextflow/cloud/google/lifesciences/GoogleLifeSciencesHelper.groovy#L245-L246

Think, it would be better to stick to the current behaviour to avoid introducing unexpected side effects. I'll do the chanege

pditommaso replied:
Even more intriguing.. actually when `entrypoint` was not specified it was defaulting to `/bin/bash` 

https://github.com/nextflow-io/nextflow/blob/96c3fd4e79ddfd028380fc0b01aef3494450c59e/plugins/nf-google/src/test/nextflow/cloud/google/lifesciences/GoogleLifeSciencesTaskHandlerTest.groovy#L185-L185

Not clear why there was that test checking for `null`

pditommaso replied:
I've removed that tests 😆

pditommaso commented on modules/nf-commons/src/main/nextflow/util/CmdLineOptionMap.groovy:
```diff
@@ -39,7 +39,7 @@ class CmdLineOptionMap implements CacheFunnel {
 
     def getFirstValue(String key) {
-        getFirstValueOrDefault(key, '')
+        getFirstValueOrDefault(key, null)
```
Usually, it's used `null` for absences of value

manuelesimi replied:
Got it. Thanks.

pditommaso commented on plugins/nf-amazon/src/main/nextflow/cloud/aws/batch/AwsContainerOptionsMapper.groovy:
```diff
@@ -47,15 +47,15 @@ class AwsContainerOptionsMapper {
             final tokens = value.tokenize('=')
             keyValuePairs << new KeyValuePair().withName(tokens[0]).withValue(tokens.size() == 2 ? tokens[1] : null)
         }
-        if ( keyValuePairs.size() > 0 )
```
This is implicit with groovy, have a look at https://groovy-lang.org/semantics.html#Groovy-Truth

manuelesimi replied:
Yep, sorry, that comes from my Java background :)

pditommaso reviewed (approved):
OK, Sunday PR approval! thanks for this, much appreciated! 

<img src="https://media3.giphy.com/media/3o6Zt6OBkOZ62CAQ4U/giphy.gif"/>

manuelesimi commented:
Terrific! This was much needed for our AWS projects.
Thanks for all the help and patience along .

pditommaso commented:
Released along with version 21.12.1-edge
